src/transport/icmp6.rs

```diff
@@ -0,0 +1,162 @@
+        // TODO ... implement the rest
+        let cksum_be = self.icmp_chksum.to_be_bytes();
+        writer.write_all(&[
+            self.icmp_type as u8,
+            self.icmp_code,
+            cksum_be[0],
+            cksum_be[1],
+        ]).map_err(WriteError::from)
+    }
+
+    pub fn calc_checksum_ipv6(&self, _ip_header: &Ipv6Header, _payload: &[u8]) -> Result<u16, ValueError> {
+        // TODO...
+        Ok(0u16)
+    }
+
+    /// Reads an icmp6 header from a slice directly and returns a tuple containing the resulting header & unused part of the slice.
+    #[inline]
+    pub fn from_slice(slice: &[u8]) -> Result<(IcmpV6Header, &[u8]), ReadError> {
+        Ok((
+            Icmp6HeaderSlice::from_slice(slice)?.to_header()
+                .map_err(|_| { ReadError::UnexpectedEndOfSlice(IcmpV6Header::SERIALIZED_SIZE)})?,
+            &slice[IcmpV6Header::SERIALIZED_SIZE..]
+        ))
+    }
+}
+
+///A slice containing an icmp6 header of a network package. Struct allows the selective read of fields in the header.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Icmp6HeaderSlice<'a> {
+    slice: &'a [u8]
+}
+
+impl<'a> Icmp6HeaderSlice<'a> {
+    /// Creates a slice containing an icmp6 header.
+    #[inline]
+    pub fn from_slice(slice: &'a[u8]) -> Result<Icmp6HeaderSlice<'a>, ReadError> {
+        //check length
+        use crate::ReadError::*;
+        if slice.len() < IcmpV6Header::SERIALIZED_SIZE {
+            return Err(UnexpectedEndOfSlice(IcmpV6Header::SERIALIZED_SIZE));
+        }
+
+        //done
+        Ok(Icmp6HeaderSlice{
+            // SAFETY:
+            // Safe as slice length is checked to be at least
+            // IcmpV6Header::SERIALIZED_SIZE (8) before this.
+            slice: unsafe {
+                from_raw_parts(
+                    slice.as_ptr(),
+                    IcmpV6Header::SERIALIZED_SIZE
+                )
+            }
+        })
+    }
+    /// Decode all the fields and copy the results to a UdpHeader struct
+    #[inline]
+    pub fn to_header(&self) -> Result<IcmpV6Header, ValueError> {
+        Ok(IcmpV6Header {
+            icmp_type: self.icmp_type()?,
+            icmp_code: self.icmp_code()?,
+            icmp_chksum: self.icmp_chksum()?,
+            echo_header: self.parse_echo_header(),
+        })
+    }
+
+    pub fn icmp_type(&self) -> Result<IcmpV6Type, ValueError> {
+        // already checked slice len in ::from_slice()
+        IcmpV6Type::from(self.slice[0])
+    }
+
+    pub fn icmp_code(&self) -> Result<u8, ValueError> {
+        // already checked slice len in ::from_slice()
+        Ok(self.slice[1])
+    }
+
+    pub fn icmp_chksum(&self) -> Result<u16, ValueError> {
+        // SAFETY:
+        // Safe as the contructor checks that the slice has
+        // at least the length of UdpHeader::SERIALIZED_SIZE (8).
+        unsafe {
+            Ok(get_unchecked_be_u16(self.slice.as_ptr().add(2)))
+        }
+    }
```

Comment:
Flag as inline, correct comment & directly return value.

```suggestion
    pub fn icmp_chksum(&self) -> u16 {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of IcmpV6Header::SERIALIZED_SIZE (8).
        unsafe {
            get_unchecked_be_u16(self.slice.as_ptr().add(2))
        }
    }
```